src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper.h

```diff
@@ -78,23 +78,36 @@ struct grpc_ares_request {
   grpc_error_handle error ABSL_GUARDED_BY(mu) = GRPC_ERROR_NONE;
 };
 
-/* Asynchronously resolve \a name. It will try to resolve grpclb SRV records in
-  addition to the normal address records. For normal address records, it uses
-  \a default_port if a port isn't designated in \a name, otherwise it uses the
-  port in \a name. grpc_ares_init() must be called at least once before this
-  function. The returned grpc_ares_request object is owned by the caller and it
-  is safe to free after on_done is called back.
+/* Asynchronously resolve \a name (A/AAAA records only).
+  It uses \a default_port if a port isn't designated in \a name, otherwise it
+  uses the port in \a name. grpc_ares_init() must be called at least once before
+  this function. The returned grpc_ares_request object is owned by the caller
+  and it is safe to free after on_done is called back.
 
   Note on synchronization: \a as on_done might be called from another thread
   ~immediately, access to the grpc_ares_request* return value must be
   synchronized by the caller. TODO(apolcyn): we should remove this requirement
   by changing this API to use two phase initialization - one API to create
   the grpc_ares_request* and another to start the async work. */
-extern grpc_ares_request* (*grpc_dns_lookup_ares)(
+extern grpc_ares_request* (*grpc_dns_lookup_hostname_ares)(
     const char* dns_server, const char* name, const char* default_port,
     grpc_pollset_set* interested_parties, grpc_closure* on_done,
     std::unique_ptr<grpc_core::ServerAddressList>* addresses,
+    int query_timeout_ms);
+
+// Asynchronously resolve a SRV record.
+// See \a grpc_dns_lookup_hostname_ares for usage details and caveats.
+extern grpc_ares_request* (*grpc_dns_lookup_srv_ares)(
```

Comment:
Agreed with the concern here, thanks for bringing it up. I went into this expecting that 3x the fds & 3x the memory are negligible given that most users will invoke a small number of short-lived, concurrent requests at any given time. I'd be interested to hear if that isn't the case.

> once we've already finished migrating the code into the EE impl, there won't be much motivation to do this refactoring

This is a common thing indeed, but since I'm doing the development on this at the moment, I don't think it's a huge risk. I subscribe to the "first: make it work; second: make it better" philosophy. Given that Alex already paved the way for an improved refactoring, I'd opt to land this change that lets us leverage the new API, and do the refactoring if we find it would be measurably meaningful (this is a flexible stance on your (1) or (3) distinction, we can adapt as needed).

That said, @apolcyn was it terribly difficult to debug the problems with the refactored resolver? I'm curious about your time investment there.

Reply:
I think I favor option 1), because this should amount to the simplest code, and at the moment I'm skeptical that the benefits of the shared state approach are worth the complexity.

In particular, I'm thinking about the following:

a) Since grpclb is deprecated, SRV queries should actually go away long term.

b) TXT queries aren't enabled by default. This point might be reconsidered once they are or are soon to be.

c) I don't know of a use case that really relies on large amounts of concurrent SRV and/or TXT queries. E.g.., personally I'd be surprised if an extra file descriptor or two per DNS resolution caused problems.

> That said, @apolcyn was it terribly difficult to debug the problems with the refactored resolver? I'm curious about your time investment there.

The problem I ran into was basically a problem with an integration test I was unfamiliar with, and which I couldn't (easily) run myself or gather more info about, to debug further. I then just didn't find time to push forward on it :(

That said, when I originally submitted that PR, it was using a work serializer. I know at least one bug fix has been made to the work serializer since then. Also, the c-ares wrapper code has since changed to use a mutex. So I wouldn't be surprised if the bug has been fixed in the interim, and maybe we could retry the test.

Reply:
As we discussed, the thing I'm concerned about here is not your individual commitment to follow up on the clean-up here, but rather whether we as a team will wind up giving you the time to do that.  What usually seems to happen to all of us is that once things are working, we get pushed toward subsequent priorities, so if we don't take time to get things right before we finish them, we never seem to actually get time to do the cleanup.

How about this: Let's go forward with this PR as-is for now, but let's fix this as part of moving the code from iomgr to the EE impl, rather than doing it as a separate clean-up step later.  Does that work?

Reply:
@markdroth by cleanup, do we mean finishing the C++ification, or by sharing the event driver between the different lookups?

If the latter, personally I'm skeptical if that's a net win, for those reasons above.

Reply:
@apolcyn I welcome you to pick up the C++ification again if you want. Otherwise, I don't think "cleanup" is necessarily the right term here, it's just a different, valid implementation. I'm still in favor of taking the flexible stance of landing this and paying attention to those who may be the most sensitive customers to determine if there's any meaningful change in performance that would necessitate a more efficient implementation.

Reply:
If nothing else, I think there will be additional memory usage here from allocating all of the state separately for each request, and given how expensive RAM is right now, that doesn't seem like a win.  Prior to this PR, we were not paying that cost, so this is making things worse, and I don't think we should do that if we have a reasonable alternative -- and I think we do.  I am willing to live with this in the short term, but I want to make sure that we will clean it up before we are finished with the EE migration.

As I said to AJ offline, I think that we are going to need to make changes to the event-loop integration anyway as part of moving this code into the EE impl, since that integration will no longer use things like pollset_set.  And since we're going to be changing that code anyway, I think centralizing it into a parent object should be reasonable to do at the same time.

As a completely separate matter, I think that if we are making a significant changes to a piece of code anyway, we should convert that code to C++ if it's still in C.  That does not necessarily mean that we need to convert every line of c-ares code, but if we're touching some given part of it anyway, I think it makes sense to convert it.  I would say the same thing about any piece of code in our codebase.

Alex, I think we should also try to learn from this for the future.  If we don't get your PR rolled forward -- and neither of you seem to want to take the time to do that, so it seems like it's not going to happen -- then the fact that we didn't follow up on it means that all the time you spent on it was basically completely wasted, because we didn't wind up getting any benefit from it.  I think we should try really hard to avoid that in the future.  If something is worth spending time on, it's worth doing it right and getting it done; if it's not, then we shouldn't spend time on it in the first place.